File: packages/shared/src/types/issue.ts
/**
 * Issue domain types for ContribHub.
 * Mirrors the API schema at apps/api/app/schemas/issue.py
 */

export enum IssueCategory {
  BUG = 'bug',
  FEATURE = 'feature',
  ENHANCEMENT = 'enhancement',
  DOCUMENTATION = 'documentation',
  QUESTION = 'question',
  SECURITY = 'security',
  PERFORMANCE = 'performance',
  TESTING = 'testing',
  REFACTOR = 'refactor',
  CHORE = 'chore',
  CI_CD = 'ci-cd',
  DESIGN = 'design',
}

export enum IssuePriority {
  CRITICAL = 'critical',
  HIGH = 'high',
  MEDIUM = 'medium',
  LOW = 'low',
}

export enum ComplexityLevel {
  TRIVIAL = 'trivial',
  EASY = 'easy',
  MODERATE = 'moderate',
  HARD = 'hard',
  EXPERT = 'expert',
}

export enum IssueStatus {
  OPEN = 'open',
  TRIAGED = 'triaged',
  ASSIGNED = 'assigned',
  IN_PROGRESS = 'in_progress',
  CLOSED = 'closed',
}

export interface Issue {
  id: string;
  repo_id: string;
  github_id: number;
  number: number;
  title: string;
  body: string | null;
  author: string;
  author_avatar_url: string | null;
  labels: string[];
  status: IssueStatus;
  category: IssueCategory | null;
  priority: IssuePriority | null;
  complexity: ComplexityLevel | null;
  category_confidence: number | null;
  priority_confidence: number | null;
  complexity_confidence: number | null;
  estimated_hours: number | null;
  required_skills: string[];
  is_good_first_issue: boolean;
  is_duplicate: boolean;
  duplicate_of_id: string | null;
  assigned_to: string | null;
  triaged_at: string | null;
  created_at: string;
  updated_at: string;
  closed_at: string | null;
}

export interface IssueCreate {
  repo_id: string;
  github_id: number;
  number: number;
  title: string;
  body: string | null;
  author: string;
  author_avatar_url: string | null;
  labels: string[];
}

export interface IssueUpdate {
  title?: string;
  body?: string | null;
  labels?: string[];
  status?: IssueStatus;
  category?: IssueCategory | null;
  priority?: IssuePriority | null;
  complexity?: ComplexityLevel | null;
  category_confidence?: number | null;
  priority_confidence?: number | null;
  complexity_confidence?: number | null;
  estimated_hours?: number | null;
  required_skills?: string[];
  is_good_first_issue?: boolean;
  is_duplicate?: boolean;
  duplicate_of_id?: string | null;
  assigned_to?: string | null;
}

export interface IssueSummary {
  id: string;
  repo_full_name: string;
  number: number;
  title: string;
  category: IssueCategory | null;
  priority: IssuePriority | null;
  complexity: ComplexityLevel | null;
  is_good_first_issue: boolean;
  created_at: string;
}

export interface IssueFilter {
  repo_id?: string;
  category?: IssueCategory;
  priority?: IssuePriority;
  complexity?: ComplexityLevel;
  status?: IssueStatus;
  is_good_first_issue?: boolean;
  labels?: string[];
  search?: string;
}

File: apps/api/app/schemas/issue.py
"""Pydantic schemas for Issues."""

from datetime import datetime
from typing import Any, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class IssueResponse(BaseModel):
    """Public representation of a triaged issue."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    repo_id: UUID
    github_id: int
    number: int
    title: str
    body: Optional[str] = None
    state: str
    author: Optional[str] = None
    labels: Optional[list[Any]] = None
    category: Optional[str] = None
    priority: Optional[str] = None
    complexity_score: Optional[int] = None
    quality_score: Optional[int] = None
    is_claimed: bool
    embedding_id: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    closed_at: Optional[datetime] = None


class IssueListParams(BaseModel):
    """Query parameters for listing issues."""

    category: Optional[str] = Field(None, pattern="^(bug|feature|question|docs|chore)$")
    priority: Optional[str] = Field(None, pattern="^(P0|P1|P2|P3)$")
    complexity_min: Optional[int] = Field(None, ge=1, le=10)
    complexity_max: Optional[int] = Field(None, ge=1, le=10)
    state: Optional[str] = Field(default="open", pattern="^(open|closed|all)$")
    is_claimed: Optional[bool] = None
    page: int = Field(default=1, ge=1)
    per_page: int = Field(default=20, ge=1, le=100)
